subprojects/diagnostics/src/main/java/org/gradle/api/tasks/diagnostics/DependencyInsightReportTask.java

```diff
@@ -393,63 +474,112 @@ private AttributeContainer concat(AttributeContainer configAttributes, Attribute
                     ((AttributeContainerInternal) dependencyAttributes).asImmutable());
         }
 
-        private void writeAttributeBlock(StyledTextOutput out, AttributeContainer attributes, AttributeContainer requested) {
-            out.withStyle(Description).text(" [");
+        private void writeAttributeBlock(StyledTextOutput out, AttributeContainer attributes, AttributeContainer requested, boolean selected) {
+            out.withStyle(Description).text("  Attributes:");
             out.println();
-            int maxAttributeLen = computeAttributePadding(attributes, requested);
-            Set<Attribute<?>> matchedAttributes = Sets.newLinkedHashSet();
-            writeFoundAttributes(out, attributes, requested, maxAttributeLen, matchedAttributes);
-            Sets.SetView<Attribute<?>> missing = Sets.difference(requested.keySet(), matchedAttributes);
-            if (!missing.isEmpty()) {
-                writeMissingAttributes(out, requested, maxAttributeLen, missing);
+            if (showSelectionReasons) {
+                writeAttributeTable(out, attributes, requested, selected);
+            } else {
+                writeFoundAttributes(out, attributes);
             }
-            out.withStyle(Description).text("   ]");
         }
 
-        private void writeMissingAttributes(StyledTextOutput out, AttributeContainer requested, int maxAttributeLen, Sets.SetView<Attribute<?>> missing) {
-            if (missing.size() != requested.keySet().size()) {
-                out.println();
+        private void writeAttributeTable(
```

Comment:
I'd break this out into multiple methods.  Divide the attributes here, then call a method to add headers, then 1 call for each type of attributes.